### ISSUE TITLE
git-filter-branch and Git LFS not working any more since 1.5.0

### ISSUE THREAD

**Issue by knime** [git-lfs/git-lfs#1773]
We used run a git-filter-branch over existing repositories in order to mark files for LFS. This worked perfectly with version 1.4.4 of Git LFS. Since 1.5.0 this doesn't work any more, the filter process fails with

```
Rewrite b5890838918144aa0922c75724778f12e41feec8 (4/588) (0 seconds passed, remaining 0 predicted)    fatal: xxx/lib/Jama-1.0.1.jar: smudge filter lfs failed
Could not checkout the index
```

This error comes when filter-branch checks out the index after the previous revision has been rewritten (Jama-1.0.1.jar has been marked as an LFS file).

**Comment by ttaylorr** [comment]
Hi, @knime, thanks for opening this and sorry for the 📻 -silence. I spent some time last week taking a look at the implementation of `git filter-branch`, and wasn't able to find a code-path that would result in this error.

Would you mind providing me with the `filter-branch` command that you ran exactly so that I can better track down the issue? If your repository has filenames that are private, feel free to name them `XXX` or similar, so long as you do it consistently.

**Comment by knime** [comment]
This simple shell script reproduces the problem:

```
git init lfs-test
cd lfs-test
echo "Test" > test.txt
git add test.txt 
git commit -m "Commit 1"
echo "Test" > test.jar
git add test.jar 
git commit -m "Commit 2"
echo "Test" > "test2.txt"
git add test2.txt 
git commit -m "Commit 3"

git lfs install --local
git filter-branch -f --prune-empty --tree-filter '
  git rm --cached -r -q .
  git lfs track "*.jar"
  git add .
' --tag-name-filter cat -- --all
```

The error message is

```
Rewrite afc881afb09c07088f0b25fdd4bdda3a72f7a573 (1/3) (0 seconds passed, remaining 0 predicted)    Tracking *.jar
Rewrite 69d6b0ce3c64620661b3d874cb31180432212b51 (2/3) (0 seconds passed, remaining 0 predicted)    fatal: test.jar: smudge filter lfs failed
Could not checkout the index
```

Git LFS version is _git-lfs/1.5.3 (GitHub; linux amd64; go 1.7.4; git 23455e4)_
Git version is _git version 2.11.0_

Version 1.4.4 works as expected.

**Comment by ttaylorr** [comment]
@knime, thanks for sending that along: it was incredibly helpful to test with locally. Turns out in working on #1796 last week, I inadvertently fixed this issue as well. Here's a little bit of background:

When running `git filter-branch` with the `--tree-filter` option, the filter gets run on each tree one-by-one (as you would expect from running `git rev-list`), but modifies state in the `.git/lfs/objects` directory since we're committing LFS objects while running through the tree(s). Here's a list of what happens:

1. Git checks out the tree at "Commit 1", and runs the `--tree-filter` command.
  a. We remove all of the files in the index with `git rm --cached -r -q .`.
  b. We tell LFS to start tracking the desired extension(s).
  c. Re-add the files to run them through LFS.
  d. Write pointer files to the index, and add object contents to `.git/lfs/objects/...`. 
2. Re-write that item in the tree, and advance the working copy to the tree at "Commit 2", which is now detached from it's previous parent "Commit 1".
3. Run the `--tree-filter` command again:
  a. Remove all of the files in the index.
  b. Have LFS track extensions (a local change, but a global no-op, since we already modified the `.gitattributes` file the last time we ran the `--tree-filter` command.
  c. Run all the LFS-tracked files through the `clean`/`smudge` or process filter.

`3c` is where the problem lies. Even though at the tree of "Commit 2", test.txt is a full text file NOT tracked by LFS, we still have a corresponding entry in `.git/lfs/objects` from the last invocation of the `--tree-filter`. This confuses LFS, which causes it to crash, and therefore fail the `filter-branch.

Prior to #1796, LFS just gave up in this case and wrote an empty file, corrupting future iterations of `filter-branch`. Now, LFS knows how to handle that case properly, and writes out a pointer file as expected.

This issue will be fixed in the v1.5.4 release which contains a backport of #1796.
